### ISSUE TITLE
creating a large raster hangs R

### ISSUE THREAD

**Issue by LDalby** [rspatial/terra#165]
Trying to create a rather big raster object throws an error and causes R to hang.

```R
library(terra)
r <- rast(ncol=35253, nrow=45153)
values(r) <- rep(1, ncell(r))
#> Error in x$.self$finalize() : attempt to apply non-function 
```
R 4.0.4, Ubuntu 20.04.2 and the latest {terra} version from GitHub

Best
Lars

**Comment by LDalby** [comment]
It does indeed @Rapsodia86! I had not seen that, thx.
However using `setValues` instead causes RStudio to crash with:
![image](https://user-images.githubusercontent.com/2366579/110523341-414ea680-8112-11eb-9872-062bcab95c58.png)

**Comment by rhijmans** [comment]
You are creating a vector of ~1.5 GB so that should be a problem --- I will look into that. I assume that you can do 

`v <- rep(1, ncell(r))`

But that this leads to the error 

`values(r) <- v`

Either way, the safe way to to this is to use `init`, as in 

`x <- init(r, 1)`

This message

`#> Error in x$.self$finalize() : attempt to apply non-function`

is not an error message in the sense that it does not stop operations (see https://github.com/rspatial/terra/issues/30)

**Comment by LDalby** [comment]
I have ~64GB RAM so no, the vector it self is not giving me problems. 

However: `x <- init(r, 1)` does actually crash the session.

**Comment by rhijmans** [comment]
I was wrong, you need about 8(bytes)*1.5(cells) = 12GB per copy --- but it should still be possible. And I can confirm that `init` seems to have a bug...

**Comment by rhijmans** [comment]
I believe that `init` now works. It might still be safest to provide a filename argument to 

```
library(terra)
r <- rast(ncol=35253, nrow=45153)
x <- init(r, 1, filename="init.tif", datatype="INT1U")
```

The below 

`values(x) <- 1` 
`values(x) <- rep(1m ncell(x))`

also work for me on ubuntu, windows and OSX; so I don't know what is going on there.  
I have changed the code such that this version 

`values(x) <- 1` 

should now be memory safe (i.e. write to disk if deemed necessary)

**Comment by LDalby** [comment]
Thanks a lot for your quick response to this issue Robert!

I now see:
```R
library(terra)  # terra version 1.1.6
r <- rast(ncol=35253, nrow=45153)
x <- init(r, 1)  # Crashes rsession and RStudio
x <- init(r, 1, filename=fs::path_home("Desktop", "init.tif"), datatype="INT1U")  # Works

values(r) <- 1  # works
values(r) <- rep(1, ncell(x))  # Works
```
So `init` without specifying a file will cause a crash. The odd thing is that I don't seem to run out of RAM. Consumption was only a about a third of the available RAM when it crashed...

**Comment by LDalby** [comment]
Just a small update here. 
Plotting `r` after having successfully assigned values to it will also result in a crash. In this case I can confirm that it ran out of RAM.

**Comment by rhijmans** [comment]
I have changed the default memory assumptions a bit. Could you try once more?

**Comment by LDalby** [comment]
Sure!

With terra version 1.1.7 all of the [above](https://github.com/rspatial/terra/issues/165#issuecomment-795005731) works.
But plotting `r` still consumes all my RAM and then RStudio crashes.

**Comment by rhijmans** [comment]
Thanks, at least some progress. The plotting crash is probably because the object gets copied at least twice before it is sampled. That is fine for a data source from disk, but not for a very large object in RAM. So apparently the values of `r` are still in memory and not in a temp file? I will change the plot code so that sampling happens first.

**Comment by LDalby** [comment]
Hmm, I wanted to see if I could confirm that the values of `r` were in memory, but now I can't seem to run the code again...
I appear to be back at square one where R hangs with the this error message `Error in x$.self$finalize() : attempt to apply non-function` 

I'm confused. When this runs on your Ubuntu setup, the problem could be on my system then 🤔 

<details style="margin-bottom:10px;">
<summary>
Session info
</summary>

``` r
sessioninfo::session_info()
#> ─ Session info ───────────────────────────────────────────────────────────────
#>  setting  value                       
#>  version  R version 4.0.4 (2021-02-15)
#>  os       Ubuntu 20.04.2 LTS          
#>  system   x86_64, linux-gnu           
#>  ui       X11                         
#>  language en_US:en                    
#>  collate  en_US.UTF-8                 
#>  ctype    en_US.UTF-8                 
#>  tz       Europe/Copenhagen           
#>  date     2021-03-16                  
#> 
#> ─ Packages ───────────────────────────────────────────────────────────────────
#>  package     * version    date       lib source                         
#>  assertthat    0.2.1      2019-03-21 [3] CRAN (R 4.0.0)                 
#>  cli           2.3.1.9000 2021-03-01 [1] local                          
#>  codetools     0.2-18     2020-11-04 [4] CRAN (R 4.0.3)                 
#>  digest        0.6.27     2020-10-24 [3] CRAN (R 4.0.3)                 
#>  evaluate      0.14       2019-05-28 [3] CRAN (R 4.0.0)                 
#>  fs            1.5.0.9000 2021-02-24 [1] local                          
#>  glue          1.4.2      2020-08-27 [3] CRAN (R 4.0.2)                 
#>  highr         0.8        2019-03-20 [3] CRAN (R 4.0.0)                 
#>  htmltools     0.5.1.1    2021-01-22 [3] CRAN (R 4.0.3)                 
#>  knitr         1.31       2021-01-27 [3] CRAN (R 4.0.3)                 
#>  lattice       0.20-41    2020-04-02 [4] CRAN (R 4.0.0)                 
#>  magrittr      2.0.1.9000 2021-02-24 [1] local                          
#>  raster        3.4-5      2020-11-14 [1] CRAN (R 4.0.4)                 
#>  Rcpp          1.0.6      2021-01-15 [3] CRAN (R 4.0.3)                 
#>  reprex        1.0.0.9002 2021-02-24 [1] local                          
#>  rlang         0.4.10     2020-12-30 [3] CRAN (R 4.0.3)                 
#>  rmarkdown     2.7        2021-02-19 [3] CRAN (R 4.0.4)                 
#>  rstudioapi    0.13       2020-11-12 [3] CRAN (R 4.0.3)                 
#>  sessioninfo   1.1.1      2018-11-05 [3] CRAN (R 4.0.0)                 
#>  sp            1.4-5      2021-01-10 [1] CRAN (R 4.0.4)                 
#>  stringi       1.5.3      2020-09-09 [3] CRAN (R 4.0.2)                 
#>  stringr       1.4.0      2019-02-10 [3] CRAN (R 4.0.0)                 
#>  terra       * 1.1-7      2021-03-16 [1] Github (rspatial/terra@3862d71)
#>  withr         2.4.1      2021-01-26 [3] CRAN (R 4.0.3)                 
#>  xfun          0.22       2021-03-11 [3] CRAN (R 4.0.4)                 
#>  yaml          2.2.1      2020-02-01 [3] CRAN (R 4.0.0)                 
#> 
#> [1] /home/au206907/R/x86_64-pc-linux-gnu-library/4.0
#> [2] /usr/local/lib/R/site-library
#> [3] /usr/lib/R/site-library
#> [4] /usr/lib/R/library
```

</details>

**Comment by rhijmans** [comment]
` Error in x$.self$finalize() : attempt to apply non-function` is a random (meaning unexplained) and annoying fluke that does not seem to affect anything (see https://github.com/rspatial/terra/issues/30)

This is what I see 

```
library(terra)
#terra version 1.1.7

#NOTE: using GDAL version 2.2.3
#For full functionality you need at least version 3.0.4

r <- rast(ncol=35253, nrow=45153)
values(r) <- rep(1, ncell(r))
terra:::.mem_info(r)

#------------------------
#Memory (GB)
#------------------------
#available       : 24.26
#allowed (60%)   : 14.56
#needed (n=1)    : 11.86
#------------------------
#proc in memory  : TRUE
#nr chunks       : 1
#------------------------


sessionInfo()
#R version 4.0.2 (2020-06-22)
#Platform: x86_64-pc-linux-gnu (64-bit)
#Running under: Ubuntu 18.04.4 LTS

#Matrix products: default
#BLAS:   /usr/lib/x86_64-linux-gnu/blas/libblas.so.3.7.1
#LAPACK: /usr/lib/x86_64-linux-gnu/lapack/liblapack.so.3.7.1

#locale:
# [1] LC_CTYPE=C.UTF-8       LC_NUMERIC=C           LC_TIME=C.UTF-8
# [4] LC_COLLATE=C.UTF-8     LC_MONETARY=C.UTF-8    LC_MESSAGES=C.UTF-8
# [7] LC_PAPER=C.UTF-8       LC_NAME=C              LC_ADDRESS=C
#[10] LC_TELEPHONE=C         LC_MEASUREMENT=C.UTF-8 LC_IDENTIFICATION=C

#attached base packages:
#[1] stats     graphics  grDevices utils     datasets  methods   base

#other attached packages:
#[1] terra_1.1-7

#loaded via a namespace (and not attached):
#[1] compiler_4.0.2   sp_1.4-5         Rcpp_1.0.6       raster_3.4-5
#[5] codetools_0.2-18 grid_4.0.2       lattice_0.20-41


```

**Comment by LDalby** [comment]
Yes, I did notice #30, but when my system throws that error it never finishes whatever process it was in. The only way I can get out of that is to kill RStudio. 

I just did a fresh install of terra and with that I can run the lines you posted just above here.

``` r
library(terra)
#> terra version 1.1.7
r <- rast(ncol=35253, nrow=45153)
values(r) <- rep(1, ncell(r)) 
terra:::.mem_info(r)
#> 
#> ------------------------
#> Memory (GB) 
#> ------------------------
#> available       : 37.21
#> allowed (60%)   : 22.33
#> needed (n=1)    : 11.86
#> ------------------------
#> proc in memory  : TRUE
#> nr chunks       : 1
#> ------------------------
```

<sup>Created on 2021-03-17 by the [reprex package](https://reprex.tidyverse.org) (v1.0.0.9002)</sup>

<details style="margin-bottom:10px;">
<summary>
Session info
</summary>

``` r
sessioninfo::session_info()
#> ─ Session info ───────────────────────────────────────────────────────────────
#>  setting  value                       
#>  version  R version 4.0.4 (2021-02-15)
#>  os       Ubuntu 20.04.2 LTS          
#>  system   x86_64, linux-gnu           
#>  ui       X11                         
#>  language en_US:en                    
#>  collate  en_US.UTF-8                 
#>  ctype    en_US.UTF-8                 
#>  tz       Europe/Copenhagen           
#>  date     2021-03-17                  
#> 
#> ─ Packages ───────────────────────────────────────────────────────────────────
#>  package     * version    date       lib source                         
#>  assertthat    0.2.1      2019-03-21 [3] CRAN (R 4.0.0)                 
#>  cli           2.3.1.9000 2021-03-01 [1] local                          
#>  codetools     0.2-18     2020-11-04 [4] CRAN (R 4.0.3)                 
#>  digest        0.6.27     2020-10-24 [3] CRAN (R 4.0.3)                 
#>  evaluate      0.14       2019-05-28 [3] CRAN (R 4.0.0)                 
#>  fs            1.5.0.9000 2021-02-24 [1] local                          
#>  glue          1.4.2      2020-08-27 [3] CRAN (R 4.0.2)                 
#>  highr         0.8        2019-03-20 [3] CRAN (R 4.0.0)                 
#>  htmltools     0.5.1.1    2021-01-22 [3] CRAN (R 4.0.3)                 
#>  knitr         1.31       2021-01-27 [3] CRAN (R 4.0.3)                 
#>  lattice       0.20-41    2020-04-02 [4] CRAN (R 4.0.0)                 
#>  magrittr      2.0.1.9000 2021-02-24 [1] local                          
#>  raster        3.4-5      2020-11-14 [1] CRAN (R 4.0.4)                 
#>  Rcpp          1.0.6      2021-01-15 [3] CRAN (R 4.0.3)                 
#>  reprex        1.0.0.9002 2021-02-24 [1] local                          
#>  rlang         0.4.10     2020-12-30 [3] CRAN (R 4.0.3)                 
#>  rmarkdown     2.7        2021-02-19 [3] CRAN (R 4.0.4)                 
#>  rstudioapi    0.13       2020-11-12 [3] CRAN (R 4.0.3)                 
#>  sessioninfo   1.1.1      2018-11-05 [3] CRAN (R 4.0.0)                 
#>  sp            1.4-5      2021-01-10 [1] CRAN (R 4.0.4)                 
#>  stringi       1.5.3      2020-09-09 [3] CRAN (R 4.0.2)                 
#>  stringr       1.4.0      2019-02-10 [3] CRAN (R 4.0.0)                 
#>  terra       * 1.1-7      2021-03-17 [1] Github (rspatial/terra@c02bcab)
#>  withr         2.4.1      2021-01-26 [3] CRAN (R 4.0.3)                 
#>  xfun          0.22       2021-03-11 [3] CRAN (R 4.0.4)                 
#>  yaml          2.2.1      2020-02-01 [3] CRAN (R 4.0.0)                 
#> 
#> [1] /home/au206907/R/x86_64-pc-linux-gnu-library/4.0
#> [2] /usr/local/lib/R/site-library
#> [3] /usr/lib/R/site-library
#> [4] /usr/lib/R/library
```

</details>

However, if I step through that very same code (as opposed to running it via `reprex::reprex()`) I get an error:
```r
library(terra)
#>terra version 1.1.7
r <- rast(ncol=35253, nrow=45153)
values(r) <- rep(1, ncell(r)) 
#> Error in (function (x)  : attempt to apply non-function
```
I have never encountered something like this before, so I am not sure how I can provide help to resolve the issue. 
If you can think of something I can do on my end, please just let me know. 

Thanks for your help and time so far!
